packages/headless/src/app/commerce-engine/commerce-engine.ts

```diff
@@ -103,6 +106,39 @@ export function buildCommerceEngine(
   return {
     ...engine,
 
+    dispatchOnSlice(slice: string) {
+      return (action: any) => {
+        return engine.dispatch({
+          ...action,
+          // Enrich the action with a slice to target the correct reducer
+          slice,
+        })
+      }
+    },
+
+    setSliceReducers(slice: string, reducers: Reducer[]) {
+      const namespaceReducer = (reducerFunction: Reducer, reducerName: string) => {
+        // TODO(nico): Handle any
+        return (state: any, action: any) => {
+          // This is where we extract the enriched sliced in `dispatchOnSlice`
+          const { slice } = action
+          const isInitializationCall = state === undefined
+          if (slice !== reducerName && !isInitializationCall) return state
+
+          return reducerFunction(state[slice], action)
+        }
+      }
+
+      const sliceReducer = combineReducers(reducers);
+
+      // TODO(nico): We're adding the reducers to the engine as top-level reducers because it's a bit difficult to get
+      //  headless to play nice with nested reducers. We should and CAN revisit this in the future since the top-level
+      //  engine state won't be exposed on the commerce engine.
+      engine.addReducers({
+        [slice]: namespaceReducer(sliceReducer, slice),
+      });
+    },
+
```

Comment:
This shows the trick used to make this approach work.

First, we create a `dispatch` specifically for namespaced actions (pagination actions, sort actions, facet actions, etc.).

Then, we expose `setSliceReducers` to allow creating the reducer for the specific instance of the solution type. This is needed because there is no way to "hook" into the reducer manager and add/remove sub-reducers. This means we must then keep the list of reducers for the solution type in the sub-controller: https://github.com/coveo/ui-kit/pull/3814/files#diff-70fb810df3b4d7646a54a4eb9c51a65a82fe48917db532f2018810b71ea340ec

The `namespaceReducer` function is adapted from https://redux.js.org/usage/structuring-reducers/reusing-reducer-logic#customizing-behavior-with-higher-order-reducers